File: legacy/python/src/core/history/analytics.py
from __future__ import annotations

from collections import defaultdict
from datetime import datetime, timedelta
from statistics import mean, pstdev
from typing import Any


def build_daily_series(price_records: list[dict[str, Any]]) -> list[dict[str, Any]]:
    grouped: dict[str, list[dict[str, Any]]] = defaultdict(list)
    for record in price_records:
        day = record["coletado_em"][:10]
        grouped[day].append(record)

    series = []
    for day in sorted(grouped):
        offers = sorted(grouped[day], key=lambda item: item["preco_total"])
        best_offer = offers[0]
        series.append(
            {
                "date": day,
                "value": float(best_offer["preco_total"]),
                "market_name": best_offer["market_name"],
                "market_id": int(best_offer["market_id"]),
                "offers": offers,
            }
        )
    return series


def compute_product_metrics(price_records: list[dict[str, Any]]) -> dict[str, Any]:
    daily_series = build_daily_series(price_records)
    if not daily_series:
        return empty_metrics()

    latest = daily_series[-1]
    latest_date = datetime.fromisoformat(latest["date"])
    values_7 = [point["value"] for point in filter_series_window(daily_series, latest_date, 7)]
    values_30 = [point["value"] for point in filter_series_window(daily_series, latest_date, 30)]
    previous_7 = [point["value"] for point in filter_series_window(daily_series[:-7], latest_date - timedelta(days=7), 7)]

    avg_7 = safe_mean(values_7)
    avg_30 = safe_mean(values_30)
    expected_low, expected_high = compute_expected_range(values_30 or values_7)
    weekly_variation = compute_variation(daily_series, 7)
    monthly_variation = compute_variation(daily_series, 30)
    trend = classify_trend(avg_7, safe_mean(previous_7))
    status = classify_status(latest["value"], expected_low, expected_high)
    minimum = min(point["value"] for point in daily_series)
    maximum = max(point["value"] for point in daily_series)

    return {
        "series": daily_series,
        "current_price": latest["value"],
        "current_market_name": latest["market_name"],
        "avg_7": avg_7,
        "avg_30": avg_30,
        "minimum": minimum,
        "maximum": maximum,
        "weekly_variation": weekly_variation,
        "monthly_variation": monthly_variation,
        "status": status,
        "trend": trend,
        "expected_range": {"low": expected_low, "high": expected_high},
        "insufficient_history_7": len(values_7) < 7,
        "insufficient_history_30": len(values_30) < 30,
    }


def compute_expected_range(values: list[float]) -> tuple[float | None, float | None]:
    if not values:
        return None, None
    center = mean(values)
    deviation = pstdev(values) if len(values) > 1 else center * 0.05
    margin = max(deviation, center * 0.05)
    return max(0.0, center - margin), center + margin


def compute_variation(series: list[dict[str, Any]], days_back: int) -> float | None:
    if len(series) < 2:
        return None

    latest_value = series[-1]["value"]
    latest_date = datetime.fromisoformat(series[-1]["date"])
    target_date = latest_date - timedelta(days=days_back)
    baseline = None

    for point in reversed(series[:-1]):
        point_date = datetime.fromisoformat(point["date"])
        if point_date <= target_date:
            baseline = point["value"]
            break

    if baseline in (None, 0):
        return None

    return ((latest_value - baseline) / baseline) * 100


def classify_status(current_price: float | None, low: float | None, high: float | None) -> dict[str, str]:
    if current_price is None or low is None or high is None:
        return {"code": "sem_dados", "label": "Sem dados"}
    if current_price < low:
        return {"code": "abaixo_faixa", "label": "Abaixo da faixa"}
    if current_price > high:
        return {"code": "acima_faixa", "label": "Acima da faixa"}
    return {"code": "dentro_faixa", "label": "Dentro da faixa"}


def classify_trend(current_avg: float | None, previous_avg: float | None) -> dict[str, str]:
    if current_avg is None or previous_avg is None or previous_avg == 0:
        return {"code": "indefinida", "label": "Historico insuficiente"}

    delta = ((current_avg - previous_avg) / previous_avg) * 100
    if delta >= 1:
        return {"code": "alta", "label": "Alta"}
    if delta <= -1:
        return {"code": "queda", "label": "Queda"}
    return {"code": "estavel", "label": "Estavel"}


def build_summary(product_name: str, metrics: dict[str, Any]) -> str:
    if metrics["current_price"] is None:
        return f"{product_name} ainda nao tem historico suficiente para um resumo confiavel."

    status = metrics["status"]["label"].lower()
    trend = metrics["trend"]["label"].lower()
    weekly_text = format_percentage(metrics["weekly_variation"])
    monthly_text = format_percentage(metrics["monthly_variation"])
    range_text = format_price_range(metrics["expected_range"]["low"], metrics["expected_range"]["high"])

    return (
        f"O menor preco atual de {product_name} esta em {format_currency(metrics['current_price'])}, "
        f"{status} para a faixa esperada de {range_text}. "
        f"A tendencia recente indica {trend}, com variacao semanal de {weekly_text} "
        f"e variacao mensal de {monthly_text}."
    )


def empty_metrics() -> dict[str, Any]:
    return {
        "series": [],
        "current_price": None,
        "current_market_name": None,
        "avg_7": None,
        "avg_30": None,
        "minimum": None,
        "maximum": None,
        "weekly_variation": None,
        "monthly_variation": None,
        "status": {"code": "sem_dados", "label": "Sem dados"},
        "trend": {"code": "indefinida", "label": "Historico insuficiente"},
        "expected_range": {"low": None, "high": None},
        "insufficient_history_7": True,
        "insufficient_history_30": True,
    }


def filter_series_window(
    series: list[dict[str, Any]],
    reference_date: datetime,
    days: int,
) -> list[dict[str, Any]]:
    threshold = reference_date - timedelta(days=days - 1)
    return [point for point in series if datetime.fromisoformat(point["date"]) >= threshold]


def safe_mean(values: list[float]) -> float | None:
    return mean(values) if values else None


def format_currency(value: float | None) -> str:
    if value is None:
        return "n/d"
    return f"R$ {value:,.2f}".replace(",", "X").replace(".", ",").replace("X", ".")


def format_percentage(value: float | None) -> str:
    if value is None:
        return "n/d"
    return f"{value:+.1f}%".replace(".", ",")


def format_price_range(low: float | None, high: float | None) -> str:
    if low is None or high is None:
        return "n/d"
    return f"{format_currency(low)} a {format_currency(high)}"

File: src/lib/domain/analytics.ts
import { formatCurrency, formatPercentage, formatPriceRange } from "@/lib/domain/format";
import type { PriceRecordWithRelations } from "@/lib/domain/types";

export type DailySeriesPoint = {
  date: string;
  value: number;
  marketName: string;
  marketId: number;
  offers: PriceRecordWithRelations[];
};

export type ProductMetrics = {
  series: DailySeriesPoint[];
  currentPrice: number | null;
  currentMarketName: string | null;
  avg7: number | null;
  avg30: number | null;
  minimum: number | null;
  maximum: number | null;
  weeklyVariation: number | null;
  monthlyVariation: number | null;
  status: { code: string; label: string };
  trend: { code: string; label: string };
  expectedRange: { low: number | null; high: number | null };
  insufficientHistory7: boolean;
  insufficientHistory30: boolean;
};

export function buildDailySeries(priceRecords: PriceRecordWithRelations[]): DailySeriesPoint[] {
  const grouped = new Map<string, PriceRecordWithRelations[]>();

  for (const record of priceRecords) {
    const date = record.collectedAt.toISOString().slice(0, 10);
    const list = grouped.get(date) ?? [];
    list.push(record);
    grouped.set(date, list);
  }

  return [...grouped.entries()]
    .sort(([left], [right]) => left.localeCompare(right))
    .map(([date, offers]) => {
      const sortedOffers = [...offers].sort((left, right) => left.totalPrice - right.totalPrice);
      const bestOffer = sortedOffers[0];

      return {
        date,
        value: bestOffer.totalPrice,
        marketName: bestOffer.marketName,
        marketId: bestOffer.marketId,
        offers: sortedOffers,
      };
    });
}

export function computeProductMetrics(priceRecords: PriceRecordWithRelations[]): ProductMetrics {
  const dailySeries = buildDailySeries(priceRecords);
  if (dailySeries.length === 0) return emptyMetrics();

  const latest = dailySeries[dailySeries.length - 1];
  const latestDate = new Date(latest.date);
  const values7 = filterSeriesWindow(dailySeries, latestDate, 7).map((point) => point.value);
  const values30 = filterSeriesWindow(dailySeries, latestDate, 30).map((point) => point.value);
  const previous7 = filterSeriesWindow(dailySeries.slice(0, -7), addDays(latestDate, -7), 7).map((point) => point.value);

  const avg7 = safeMean(values7);
  const avg30 = safeMean(values30);
  const [low, high] = computeExpectedRange(values30.length > 0 ? values30 : values7);

  return {
    series: dailySeries,
    currentPrice: latest.value,
    currentMarketName: latest.marketName,
    avg7,
    avg30,
    minimum: Math.min(...dailySeries.map((point) => point.value)),
    maximum: Math.max(...dailySeries.map((point) => point.value)),
    weeklyVariation: computeVariation(dailySeries, 7),
    monthlyVariation: computeVariation(dailySeries, 30),
    status: classifyStatus(latest.value, low, high),
    trend: classifyTrend(avg7, safeMean(previous7)),
    expectedRange: { low, high },
    insufficientHistory7: values7.length < 7,
    insufficientHistory30: values30.length < 30,
  };
}

export function computeExpectedRange(values: number[]): [number | null, number | null] {
  if (values.length === 0) return [null, null];
  const center = safeMean(values) ?? 0;
  const deviation = values.length > 1 ? populationStdDev(values) : center * 0.05;
  const margin = Math.max(deviation, center * 0.05);
  return [Math.max(0, center - margin), center + margin];
}

export function computeVariation(series: DailySeriesPoint[], daysBack: number): number | null {
  if (series.length < 2) return null;

  const latest = series[series.length - 1];
  const latestDate = new Date(latest.date);
  const targetDate = addDays(latestDate, -daysBack);
  const baseline = [...series.slice(0, -1)].reverse().find((point) => new Date(point.date) <= targetDate)?.value;

  if (baseline == null || baseline === 0) return null;
  return ((latest.value - baseline) / baseline) * 100;
}

export function classifyStatus(currentPrice: number | null, low: number | null, high: number | null) {
  if (currentPrice == null || low == null || high == null) return { code: "sem_dados", label: "Sem dados" };
  if (currentPrice < low) return { code: "abaixo_faixa", label: "Abaixo da faixa" };
  if (currentPrice > high) return { code: "acima_faixa", label: "Acima da faixa" };
  return { code: "dentro_faixa", label: "Dentro da faixa" };
}

export function classifyTrend(currentAvg: number | null, previousAvg: number | null) {
  if (currentAvg == null || previousAvg == null || previousAvg === 0) {
    return { code: "indefinida", label: "Historico insuficiente" };
  }

  const delta = ((currentAvg - previousAvg) / previousAvg) * 100;
  if (delta >= 1) return { code: "alta", label: "Alta" };
  if (delta <= -1) return { code: "queda", label: "Queda" };
  return { code: "estavel", label: "Estavel" };
}

export function buildSummary(productName: string, metrics: ProductMetrics): string {
  if (metrics.currentPrice == null) {
    return `${productName} ainda nao tem historico suficiente para uma leitura analitica confiavel.`;
  }

  const statusText =
    metrics.status.code === "abaixo_faixa"
      ? "abaixo da faixa esperada, sugerindo uma janela favoravel de compra"
      : metrics.status.code === "acima_faixa"
        ? "acima da faixa esperada, indicando pressao de preco no curto prazo"
        : "dentro da faixa esperada, em um nivel considerado saudavel para acompanhamento";

  const trendText =
    metrics.trend.code === "alta"
      ? "a tendencia recente segue de alta"
      : metrics.trend.code === "queda"
        ? "a tendencia recente aponta acomodacao"
        : "a tendencia recente permanece estavel";

  return [
    `${productName} registra menor preco atual de ${formatCurrency(metrics.currentPrice)},`,
    `${statusText}.`,
    `${trendText}, com variacao semanal de ${formatPercentage(metrics.weeklyVariation)}`,
    `e variacao mensal de ${formatPercentage(metrics.monthlyVariation)}.`,
    `A faixa esperada neste recorte vai de ${formatPriceRange(metrics.expectedRange.low, metrics.expectedRange.high)}.`,
  ].join(" ");
}

export function emptyMetrics(): ProductMetrics {
  return {
    series: [],
    currentPrice: null,
    currentMarketName: null,
    avg7: null,
    avg30: null,
    minimum: null,
    maximum: null,
    weeklyVariation: null,
    monthlyVariation: null,
    status: { code: "sem_dados", label: "Sem dados" },
    trend: { code: "indefinida", label: "Historico insuficiente" },
    expectedRange: { low: null, high: null },
    insufficientHistory7: true,
    insufficientHistory30: true,
  };
}

function filterSeriesWindow(series: DailySeriesPoint[], referenceDate: Date, days: number) {
  const threshold = addDays(referenceDate, -(days - 1));
  return series.filter((point) => new Date(point.date) >= threshold);
}

function safeMean(values: number[]): number | null {
  if (values.length === 0) return null;
  return values.reduce((sum, value) => sum + value, 0) / values.length;
}

function populationStdDev(values: number[]): number {
  const avg = safeMean(values) ?? 0;
  const variance = values.reduce((sum, value) => sum + (value - avg) ** 2, 0) / values.length;
  return Math.sqrt(variance);
}

function addDays(date: Date, days: number) {
  const next = new Date(date);
  next.setDate(next.getDate() + days);
  return next;
}
